lib/diagnostics_channel.js

```diff
@@ -136,10 +138,136 @@ function hasSubscribers(name) {
   return channel.hasSubscribers;
 }
 
+class TracingChannel {
+  constructor(name) {
+    this.name = name;
+    this.channels = {
+      start: channel(`${name}.start`),
+      end: channel(`${name}.end`),
+      asyncEnd: channel(`${name}.asyncEnd`),
+      error: channel(`${name}.error`)
+    };
+  }
+
+  get hasSubscribers() {
+    const { channels } = this;
+    for (const key in channels) {
+      if (channels[key].hasSubscribers) return true;
+    }
+    return false;
+  }
+
+  subscribe(handlers) {
+    let subscribed = true;
+    for (const key in handlers) {
+      if (!subscribe(`${this.name}.${key}`, handlers[key])) {
+        subscribed = false;
+      }
+    }
+    return subscribed;
+  }
+
+  unsubscribe(handlers) {
+    let unsubscribed = true;
+    for (const key in handlers) {
+      if (!unsubscribe(`${this.name}.${key}`, handlers[key])) {
+        unsubscribed = false;
+      }
+    }
+    return unsubscribed;
+  }
+
+  traceSync(fn, ctx = {}, thisArg, ...args) {
+    const { start, end, error } = this.channels;
+    start.publish(ctx);
+    try {
+      const result = ReflectApply(fn, thisArg, args);
+      ctx.result = result;
+      return result;
+    } catch (err) {
+      ctx.error = err;
+      error.publish(ctx);
+      throw err;
+    } finally {
+      end.publish(ctx);
+    }
+  }
+
+  tracePromise(fn, ctx = {}, thisArg, ...args) {
+    const { asyncEnd, start, end, error } = this.channels;
+    start.publish(ctx);
+
+    const reject = (err) => {
+      ctx.error = err;
+      error.publish(ctx);
+      asyncEnd.publish(ctx);
+      throw err;
+    };
+
+    const resolve = (result) => {
+      ctx.result = result;
+      asyncEnd.publish(ctx);
+      return result;
+    };
+
+    try {
+      const promise = ReflectApply(fn, thisArg, args);
```

Comment:
That's fine. Error rejections are handled via line 202.

Reply:
I mean that the try-catch is not needed or anyway it could implemented differently. If `fn` does not return a promise, the line afterward will fail.

Reply:
I would _like_ to agree, but there are plenty of examples of people writing promise-returning functions that are able to throw rather than reject because they didn't use async functions and didn't wrap the whole body in a `new Promise(...)`. Better to be safe here. 😅